Add a CORS builder

markstory opened the pull request:
This adds a CORS builder that can be used to set the various cors headers through a fluent interface. This was originally discussed in #7402 and #7405.

I've not finished the documentation, but I wanted to get feedback on these changes before writing all the documentation.
### Todo
- [x] Finish documentation.
- [ ] Update book docs.

markstory committed:
Start building out the CorsBuilder.

This class will provide a fluent interface for defining all the
Access-Control-* headers. The current implementation only adds headers
if there is an origin match. I'm going to address that issue once more
of the methods are in place.

markstory committed:
Implement more methods on CorsBuilder.

markstory committed:
Finish building out methods for CorsBuilder.

markstory committed:
Integrate CorsBuilder into Response.

This maintains compatiblity with previous releases, and also allows the
builder to be used for addtional options.

markstory committed:
Deprecate the parameters for Response::cors().

The builder gives a more useful and expressive interface for setting
cors options. The other parameters will be removed in 4.x

lorenzo commented:
I'm a bit concerned about how this new builder would combine with the PSR-7 changes, since the builder is holding a reference to the response object.

markstory commented:
@lorenzo A valid concern. Would it be better if at the end of the setters, there was an `apply()` or `build()` method that returned the modified/new response?

lorenzo commented:
@markstory I think that could work nicely

markstory commented:
@lorenzo Ok, I'll update the code to have an `build()` method as most builders I've used implement that method and not `apply()`.

markstory committed:
Add a build() method to CorsBuilder.

Having a build() method lets us do 2 things:

* Call the cors methods in any order.
* Still enforce the constraint that headers should only be set for
  when an allowed origin is set.

It also makes the CorsBuilder easier to integrate in the future when we
support PSR7 request/response object.

markstory committed:
Update Response test.

header() is only called once now, and using a mock doesn't make a lot of
sense anymore.

markstory commented:
Build method done now.

lorenzo commented:
@markstory looks good :)

markstory committed:
Add doc blocks for CorsBuilder.

markstory commented:
Doc blocks added.

ADmad commented:
Few CS errors reported for `CorsBuilder.php`

markstory committed:
Fix PHPCS errors.

ADmad commented:
:+1: CS errors in other files are unrelated and already fixed.

lorenzo commented:
Nice job!

markstory commented:
I'll add some book docs tonight.
